messaging/src/main/java/org/axonframework/eventhandling/replay/ReplayContextParameterResolverFactory.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2010-2022. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventhandling.replay;
+
+import org.axonframework.eventhandling.ReplayToken;
+import org.axonframework.messaging.Message;
+import org.axonframework.messaging.annotation.ParameterResolver;
+import org.axonframework.messaging.annotation.ParameterResolverFactory;
+
+import java.lang.reflect.Executable;
+import java.lang.reflect.Parameter;
+
+/**
+ * An implementation of the {@link ParameterResolverFactory} which resolves the parameter annotated with
+ * {@link ReplayContext}. Will resolve the parameter if the {@link Message} is a
+ * {@link org.axonframework.eventhandling.TrackedEventMessage}, containing a {@link ReplayToken} with a matching context
+ * of that type. Otherwise, it will resolve always to null.
+ * <p>
+ * This parameter resolver will always match to prevent missing event handlers.
+ *
+ * @author Mitchell Herrijgers
+ * @since 4.6.0
+ */
+public class ReplayContextParameterResolverFactory implements ParameterResolverFactory {
+
+    @Override
+    public ParameterResolver createInstance(Executable executable, Parameter[] parameters, int parameterIndex) {
+        Parameter parameter = parameters[parameterIndex];
+        if (parameter.isAnnotationPresent(ReplayContext.class)) {
+            return new ReplayParameterResolver(parameter.getType());
+        }
+        return null;
+    }
+
+    private static class ReplayParameterResolver implements ParameterResolver<Object> {
+
+        private final Class<?> type;
+
+        public ReplayParameterResolver(Class<?> type) {
+            this.type = type;
+        }
+
+        @Override
+        public Object resolveParameterValue(Message message) {
+            return ReplayToken.replayContext(message, this.type).orElse(null);
+        }
+
+        @Override
+        public boolean matches(Message message) {
+            return true;
```

Comment:
I'd at least expect a validation that the `Message` is an instance of `TrackedEventMessage`. Retrieving a reset/replay context only makes sense in the context of a Streaming Event Processing.

Differently put, I don't think it makes sense that a Command Handler or Query Handler contains a parameter annotated with `ReplayContext` is still invoked.
Or, that a `SubscribingEventProcessor` behind an Event Handling Component containing the `ReplayContext` annotation on event handler parameters is still invoked.

Reply:
Fair enough, changing this